examples/keras_recipes/approximating_non_function_mappings.py

```diff
@@ -0,0 +1,424 @@
+
+Neural networks are universal function approximators. Key word: function!
+While powerful function approximators, neural networks are not able to
+approximate non-functions.
+One important restriction to remember about functions - they have one input, one
+output!
+Neural networks suffer greatly when the training set has multiple values of Y for a single X.
+
+In this guide I'll show you how to approximate the class of non-functions
+consisting of mappings from `x -> y` such that multiple `y` may exist for a
+given `x`.  We'll use a class of neural networks called
+"Mixture Density Networks".
+
+I'm going to use the new
+[multibackend Keras Core project](https://github.com/keras-team/keras-core) to
+build my Mixture Density networks.
+Great job to the Keras team on the project - it's awesome to be able to swap
+frameworks in one line of code.
+
+Some bad news: I use TensorFlow probability in this guide... so it doesn't
+actually work with other backends.
+
+Anyways, let's start by installing dependencies and sorting out imports:
+"""
+"""shell
+pip install -q --upgrade tensorflow-probability keras-core
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+from collections import OrderedDict
+import math
+import random
+from keras_core import callbacks
+import keras_core
+import tensorflow as tf
+from keras_core import layers
+from keras_core import optimizers
+from tensorflow_probability import distributions as tfd
+
+"""
+Next, lets generate a noisy spiral that we're going to attempt to approximate.
+I've defined a few functions below to do this:
+"""
+
+
+def normalize(x):
+    return (x - np.min(x)) / (np.max(x) - np.min(x))
+
+
+def create_noisy_spiral(n, max_jitter=0.2, size_range=[5, 10]):
+    R_X = random.choice(range(size_range[0], size_range[1], 10))
+    R_Y = random.choice(range(size_range[0], size_range[1], 10))
+    X_0 = random.random() * 100 - 50
+    Y_0 = random.random() * 100 - 50
+
+    randomized_jitter = lambda: random.uniform(-max_jitter / 2, max_jitter / 2)
+    x = lambda _t: _t / 360.0 * math.cos(_t / 90.0 * math.pi) * R_X + X_0
+    y = lambda _t: _t / 360.0 * math.sin(_t / 90.0 * math.pi) * R_Y + Y_0
+
+    out = np.zeros([n, 2])
+
+    for i in range(n):
+        t = 360.0 / n * i
+        out[i, 0] = x(t) + randomized_jitter()
+        out[i, 1] = y(t) + randomized_jitter()
+    out = out.astype("float32")
+    return (normalize(out[:, 0]) * 10) - 5, (normalize(out[:, 1]) * 10) - 5
+
+
+"""
+Next, lets invoke this function many times to construct a sample dataset:
+"""
+
+xs = []
+ys = []
+
+for _ in range(10):
+    x, y = create_noisy_spiral(1000)
+    xs.append(x)
+    ys.append(y)
+
+x = np.concatenate(xs, axis=0)
+y = np.concatenate(ys, axis=0)
+x = np.expand_dims(x, axis=1)
+
+plt.scatter(x, y)
+plt.show()
+
+"""
+As you can see, there's multiple possible values for Y with respect to a given
+X.  Normal neural networks will simply learn the mean of these points with
+respect to geometric space.
```

Comment:
However, in the context of our spiral the geometric mean of the each Y occurs with a probability of zero...